File: btap/Baitap_1/app/src/main/java/com/example/baitap_1/MainActivity.java
package com.example.baitap_1;


import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity {

    private TextView statusText;
    private ExecutorService executorService;
    private Handler mainHandler;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        statusText = findViewById(R.id.statusText);
        executorService = Executors.newSingleThreadExecutor();
        mainHandler = new Handler(Looper.getMainLooper());

        if (isNetworkAvailable()) {
            new CheckInternetTask(executorService, mainHandler, statusText).execute("https://www.google.com");
        } else {
            statusText.setText(getString(R.string.no_network));
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (executorService != null) {
            executorService.shutdown();
        }
    }

    private boolean isNetworkAvailable() {
        ConnectivityManager cm =
                (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);

        if (cm == null) {
            return false;
        }

        Network network = cm.getActiveNetwork();
        if (network == null) {
            return false;
        }

        NetworkCapabilities capabilities = cm.getNetworkCapabilities(network);
        return capabilities != null &&
                (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
                 capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) ||
                 capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET));
    }

    private static class CheckInternetTask {
        private final ExecutorService executor;
        private final Handler mainHandler;
        private final TextView statusText;

        CheckInternetTask(ExecutorService executor, Handler mainHandler, TextView statusText) {
            this.executor = executor;
            this.mainHandler = mainHandler;
            this.statusText = statusText;
        }

        void execute(String url) {
            executor.execute(() -> {
                boolean result = checkConnection(url);
                mainHandler.post(() -> {
                    if (result) {
                        statusText.setText(statusText.getContext().getString(R.string.connected_success));
                    } else {
                        statusText.setText(statusText.getContext().getString(R.string.cannot_connect));
                    }
                });
            });
        }

        private boolean checkConnection(String urlString) {
            try {
                HttpURLConnection urlConnection =
                        (HttpURLConnection) new URL(urlString).openConnection();

                urlConnection.setRequestMethod("HEAD");
                urlConnection.setConnectTimeout(3000);
                urlConnection.setReadTimeout(3000);
                urlConnection.connect();

                int responseCode = urlConnection.getResponseCode();
                return (responseCode == 200);

            } catch (IOException e) {
                return false;
            }
        }
    }
}

File: MyApplication/app/src/main/java/com/example/myapplication/MainActivity.kt
package com.example.myapplication

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.HttpURLConnection
import java.net.URL

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContent {
            MaterialTheme {
                NetworkCheckScreen()
            }
        }
    }
}

@Composable
fun NetworkCheckScreen() {
    val context = LocalContext.current

    var statusText by remember { mutableStateOf("Đang kiểm tra...") }

    // chạy kiểm tra 1 lần khi mở app
    LaunchedEffect(Unit) {
        if (isNetworkAvailable(context)) {
            val ok = checkInternet("https://www.google.com")
            statusText = if (ok) {
                "✅ Kết nối Internet thành công (Google OK)"
            } else {
                "⚠️ Có mạng nhưng không truy cập được Google"
            }
        } else {
            statusText = "❌ Không có kết nối mạng"
        }
    }

    // UI
    Box(
        modifier = Modifier.fillMaxSize(),
        contentAlignment = Alignment.Center
    ) {
        Text(text = statusText)
    }
}

// Kiểm tra có mạng không (WiFi / Mobile Data)
fun isNetworkAvailable(context: Context): Boolean {
    val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    val network = cm.activeNetwork ?: return false
    val capabilities = cm.getNetworkCapabilities(network) ?: return false

    return capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) ||
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)
}

// Check Internet bằng cách kết nối Google
suspend fun checkInternet(urlString: String): Boolean {
    return withContext(Dispatchers.IO) {
        try {
            val urlConnection = URL(urlString).openConnection() as HttpURLConnection
            urlConnection.requestMethod = "HEAD"
            urlConnection.connectTimeout = 3000
            urlConnection.readTimeout = 3000
            urlConnection.connect()

            val responseCode = urlConnection.responseCode
            responseCode == 200
        } catch (e: Exception) {
            false
        }
    }
}
